lib/shim/message-shim.js

```diff
@@ -361,34 +361,67 @@ function recordConsume(nodule, properties, spec) {
     if (shim.isFunction(spec)) {
       spec = spec.apply(this, arguments)
     } else {
-      spec = { ...DEFAULT_SPEC, ...spec }
+      spec = shim.setDefaults(spec, DEFAULT_SPEC)
       const destIdx = shim.normalizeIndex(args.length, spec.destinationName)
       if (destIdx !== null) {
         spec.destinationName = args[destIdx]
       }
     }
     const messageHandler = spec.messageHandler
-    const getParams = shim.agent.config.message_tracer.segment_parameters.enabled
 
     if (messageHandler) {
-      spec.after = function afterPromise({ shim, val, segment }) {
-        const msgProps = messageHandler.call(this, shim, fn, fnName, val)
-        if (getParams && msgProps?.parameters) {
-          shim.copySegmentParameters(segment, msgProps.parameters)
-        }
-        return val
-      }
+      _wrapCallBackWithMessageHandler(shim, messageHandler, args, spec)
     }
 
     const name = _nameMessageSegment(shim, spec, shim._metrics.CONSUME)
     return {
-      name: name,
+      name,
       recorder: genericRecorder,
       ...spec
     }
   })
 }
 
+/**
+ * Helper function for recordConsume
+ *
+ * @private
+ * @param {MessageShim} shim
+ *  The shim used to wrap the subscribe function
+ * @param {Function} messageHandler
+ *  The user-supplied message handler
+ * @param {Array} args
+ *  The arguments supplied to the message handler
+ * @param {MessageSubscribeSpec} spec
+ *  The user-supplied specification for this recording method's interface.
+ */
+function _wrapCallBackWithMessageHandler(shim, messageHandler, args, spec) {
+  const getParams = shim.agent.config.message_tracer.segment_parameters.enabled
+  let cbArgs
+  let origCb
+  let origCbName
+  const cbIdx = shim.normalizeIndex(args.length, spec.callback)
+  if (cbIdx !== null) {
+    // wrap the callback to save the args
```

Comment:
this is hard to follow but comparing from before, it was doing shim.bindCallbackSegment. do we no longer need this because it's using shim.record above?  Also the wrapped callback is confusing to me, can you explain why it's only storing the original args as a reference then calling the original callback?  Also, not understanding the defineProperty of name to keep function the same

Reply:
Okay, yeah, so maybe this whole thing didn't actually reduce cognitive complexity. Here's the situation:

We have to wrap a function (let's call it `getMessage`, like in the `newrelic-examples` case) that calls a callback (let's call it `callback`). The old method was doing this by calling `shim.wrap` inside a `shim.wrap`. That wraps both `getMessage` and `callback`. This works but it ends up duplicating a lot of the functionality of `spec.record` and is complicated to follow.

But new method now calls `shim.record` and uses `spec.after` to call `getMessage`'s callback. This almost works. The callback gets called by `spec.after` at the right time. The only problem is that `spec.after`, like its name implies, is called *after* `callback`. So first `callback` is called and then `spec.after` is called which in this case does stuff like putting the parameters into the segment. Defining the appropriate segments and their attributes for both `getMessage` and `callback` is now all handled by the `shim.record` function.

The problem is that because `spec.after` is called after, we don't have `callback`'s arguments. Those were invoked when `getMessage` called `callback`.

So I patched this by doing a minimal, ad-hoc wrapping of `callback` that all it does is grab its arguments, and not all of `shim.wrap`'s other things. Because I don't want to change the way segments are named (that would break a test in `test/versioned/amqplib/callback.tap.js`), I also make sure the wrapped callback has the same function name as `callback`.

What do you think? Maybe the whole thing is now just simply complicated in a different way and this was no improvement at all.

Reply:
this is not working as you expect.  the wrapped callback gets called after the .after function below.  this is because of ordering in event loop.  I also don't think this is the right way to approach this. I'm still digging into the _right_ way but it's getting very confusing 😄 .  Honestly at this point I'm wondering if we just keep this open and go forward with the major release.